src/pg_query_summary.c

```diff
@@ -0,0 +1,775 @@
+
+		case T_RawStmt:
+			return pg_query_summary_walk_impl(castNode(RawStmt, node)->stmt, state);
+
+		case T_FuncCall:
+			add_function(castNode(FuncCall, node)->funcname, CONTEXT_CALL, state);
+			break;
+
+		// The following are added to avoid needing subselect_items,
+		// like is used in Rust and Ruby.
+
+		case T_ColumnRef: {
+			if (!state->save_filter_columns)
+				break;
+
+			ColumnRef* ref = castNode(ColumnRef, node);
+
+			Node* column_node = llast(ref->fields);
+
+			// Check that we are not in the A_Star case here.
+			// (Columns can be stars, but nothing else can.)
+			if (!IsA(column_node, String))
+				break;
+
+			String* schema_name = NULL;
+			String* table_name = NULL;
+			String* column = castNode(String, column_node);
+
+			size_t num_fields = list_length(ref->fields);
+
+			if (num_fields == 2) {
+				table_name = linitial(ref->fields);
+			} else if (num_fields == 3) {
+				schema_name = linitial(ref->fields);
+				table_name = castNode(String, list_nth_cell(ref->fields, 1));
+			}
+
+			FilterColumn* fc = palloc(sizeof(FilterColumn));
+			fc->schema_name = schema_name ? schema_name->sval : NULL;
+			fc->table_name = table_name ? table_name->sval : NULL;
+			fc->column = column ? column->sval : NULL;
+
+			state->filter_columns = lappend(state->filter_columns, fc);
+
+			break;
+		}
+
+		// Things which we are only handling because raw_expression_tree_walker()
+		// doesn't handle them.
+
+		case T_CheckPointStmt:
+		case T_CreateSchemaStmt:
+		case T_GrantRoleStmt:
+		case T_TransactionStmt:
+		case T_VariableShowStmt:
+		case T_VariableSetStmt:
+			return false;
+
+		default:
+			break;
+	}
+
+	return raw_expression_tree_walker(node, pg_query_summary_walk_impl, (void*) state);
+}
+
+static char* range_var_parts_to_name(char* schemaname, char* relname) {
+	StringInfoData buf;
+	initStringInfo(&buf);
+	if (schemaname != NULL) {
+		appendStringInfoString(&buf, quote_identifier(schemaname));
+		appendStringInfoChar(&buf, '.');
+	}
+	appendStringInfoString(&buf, quote_identifier(relname));
+	return buf.data;
+}
+
+// Walk through all collected cte_names, and compares them to relname.
+//
+// Returns true if any are an exact match, false otherwise.
+static bool cte_names_include(WalkState* state, char* relname) {
+	ListCell* lc = NULL;
+	foreach(lc, state->cte_names) {
+		char *ctename = lfirst(lc);
+		if (strcmp(ctename, relname) == 0)
+			return true;
+	}
+	return false;
+}
+
+static void handle_from_clause_item(Node* node, ContextType context, WalkState* state) {
+	if (node == NULL)
+		return;
+
+	switch (nodeTag(node)) {
+		case T_JoinExpr: {
+			JoinExpr* je = castNode(JoinExpr, node);
+
+			handle_from_clause_item((Node*) je->larg, context, state);
+			handle_from_clause_item((Node*) je->rarg, context, state);
+
+			handle_from_clause_item(je->quals, context, state);
+
+			break;
+		}
+		case T_RangeVar: {
+			RangeVar* rv = castNode(RangeVar, node);
+			add_table(state, context, castNode(RangeVar, node));
```

Comment:
With this method simplified to just handle RangeVars we could inline `add_table` here for clarity and to discourage anyone from calling it directly.